File: matrix/matrix.py
from typing import Any

class MyMatrix:
    '''
    implement matrix by python
    '''

    def __init__(self, row: int) -> None:
        self.matrix: list = list()
        for i in range(row):
            self.matrix.append([])

    def add(self, row: int, value: Any) -> bool:
        '''
        add an element to a given row
        '''
        if row > len(self.matrix):
            print("row number is larger than matrix row length")
            return False
        else:
            self.matrix[row].append(value)

    def modify(self) -> None:
        '''
        modify to a matrix
        '''
        maxRowLength = 0
        for i in range(len(self.matrix)):
            if maxRowLength < len(self.matrix[i]):
                maxRowLength = len(self.matrix[i])
        for i in range(len(self.matrix)):
            for j in range(maxRowLength - len(self.matrix[i])):
                self.matrix[i].append(0);

    def shape(self) -> list:
        '''
        matrix's shape
        '''
        row = len(self.matrix)
        column = len(self.matrix[0])
        return [row, column]
    
    def print(self) -> None:
        '''
        print all elements
        '''
        for i in range(len(self.matrix)):
            for j in range(len(self.matrix[i])):
                print(self.matrix[i][j])
    
    def print_by_row(self) -> None:
        '''
        print elements by row
        '''
        for i in range(len(self.matrix)):
            print("$i row:")
            for j in range(len(self.matrix[i])):
                print(self.matrix[i][j])

    def print_by_column(self) -> None:
        '''
        print elements by column
        '''
        for i in range(len(self.matrix[0])):
            print("$i column:")
            for j in range(len(self.matrix)):
                print(self.matrix[j][i])

    def transpose(self) -> None:
        '''
        transpose a matrix
        '''
        newColNum = self.shape()[0]
        newRowNum = self.shape()[1]
        newRow = list()
        for i in range(len(self.matrix[0])):
            eachRow = list()
            for j in range(len(self.matrix)):
                eachRow.append(self.matrix[j][i])
            newRow.append(eachRow)
        for i in range(len(newRow)):
            self.matrix.append(newRow[i])
        for i in range(newColNum):
            self.matrix.pop(0)

File: matrix/matrix.js
// implement matrix by ES6 class syntax
class MyMatrix {
  constructor(row) {
    this.matrix = [];
    for (let i = 0; i < row; i += 1) {
      this.matrix.push([]);
    }
  }
  
  // add an element to a given row
  add(row, value) {
    if (row > this.matrix.length) {
      console.log('row number is larger than matrix row length');
      return false;
    } else {
      this.matrix[row].push(value);
    }
  }
  
  // modify to a matrix
  modify() {
    let maxRowLength = 0;
    for (let i = 0; i < this.matrix.length; i += 1) {
      if (maxRowLength < this.matrix[i].length) {
        maxRowLength = this.matrix[i].length;
      }
    }
    for (let i = 0; i < this.matrix.length; i += 1) {
      for (let j = 0; j < maxRowLength - this.matrix[i].length; j += 1) {
        this.matrix[i].push(0);
      }
    }
  }

  // matrix's shape
  shape() {
    const row = this.matrix.length;
    const column = this.matrix[0].length;
    return { row, column };
  }

  // print all elements
  print() {
    this.matrix.map((item) => {
      item.map((element) => {
        console.log(element);
      });
    });
  }

  // print elements by row
  printByRow() {
    for (let i = 0; i < this.matrix.length; i += 1) {
      console.log(`${i} row:'`);
      for (let j = 0; j < this.matrix[i].length; j += 1) {
        console.log(this.matrix[i][j]);
      }
    }
  }

  // print elements by column
  printByColumn() {
    for (let i = 0; i < this.matrix[0].length; i += 1) {
      console.log(`${i} column:`);
      for (let j = 0; j < this.matrix.length; j += 1) {
        console.log(this.matrix[j][i]);
      }
    }
  }

  // transpose a matrix
  transpose() {
    const { newColNum, newRowNum } = this.shape();
    let newRow = [];
    for (let i = 0; i < this.matrix[0].length; i += 1) {
      let eachRow = [];
      for (let j = 0; j < this.matrix.length; j += 1) {
        eachRow.push(this.matrix[j][i]);
      }
      newRow.push(eachRow);
    } 
    for (let i = 0; i < newRow.length; i += 1) {
      this.matrix.push(newRow[i]);
    }
    for (let i = 0; i < newColNum; i += 1) {
      this.matrix.shift();
    }
  }
}

